bee-message/src/output/feature_block/mod.rs

```diff
@@ -293,25 +308,50 @@ fn validate_dependencies(feature_blocks: &[FeatureBlock]) -> Result<(), Error> {
 
 pub(crate) fn validate_allowed_feature_blocks(
     feature_blocks: &FeatureBlocks,
-    allowed_feature_blocks_kind: &[u8],
+    allowed_feature_blocks: FeatureBlockUsages,
 ) -> Result<(), Error> {
-    if feature_blocks.len() > allowed_feature_blocks_kind.len() {
-        return Err(Error::TooManyFeatureBlocks {
-            max: allowed_feature_blocks_kind.len(),
-            actual: feature_blocks.len(),
-        });
-    }
-
-    let mut index = 0;
-
-    for feature_block in feature_blocks.iter() {
-        index = allowed_feature_blocks_kind[index..]
-            .binary_search(&feature_block.kind())
-            .map_err(|index| Error::UnallowedFeatureBlock {
+    let mut blocks = FeatureBlockUsages::empty();
+    for (index, feature_block) in feature_blocks.iter().enumerate() {
+        let usage = feature_block.usage();
+        if blocks.contains(usage) {
+            return Err(Error::DuplicateFeatureBlock {
```

Comment:
You don't need to check for duplicates, an already constructed `FeatureBlocks` ensures that there are no duplicates.

Reply:
Even better 😄